spec/frameSpec.js

```diff
@@ -0,0 +1,162 @@
+describe("Frame", function() {
+  var Frame = require("../lib/frame")
+  var Game = require("../lib/game")
+  var frame;
+  var frameTwo;
+  var finalFrame;
+  var game;
+
+  beforeEach(function() {
+    game = new Game;
+    frame = game.frames[0];
+    frameTwo = game.frames[1];
+    finalFrame = game.frames[9];
+  });
+
+  describe("total", function() {
+    it("starts at 0", function() {
+      expect(frame.total).toEqual(0);
+    });
+  });
+
+  describe(".add()", function() {
+    it("adds value to total", function() {
+      frame.add(5);
+      expect(frame.total).toEqual(5);
+    });
+  });
+
+  describe(".roll()", function() {
+    it("adds roll to rolls", function() {
+      frame.roll(5);
+      expect(frame.rolls).toEqual(1);
+    });
+    it("adds its value to previous frame totals if they are due a bonus", function() {
+      frame.roll(10);
+      frameTwo.roll(5);
+      expect(frame.total).toEqual(15);
+    });
+  });
+
+  describe(".hasBonus()", function() {
+    it("begins false", function() {
+      expect(frame.hasBonus()).toEqual(false);
+    });
+  });
+
+  describe(".bonusRolls", function() {
+    it("begins empty", function() {
+      expect(frame.bonusRolls).toEqual(0);
+    });
+    it("increments by 2 after strike", function() {
+      frame.roll(10);
+      expect(frame.bonusRolls).toEqual(2);
+    });
+    it("increments by 1 after spare", function() {
+      frame.roll(5);
+      frame.roll(5);
+      expect(frame.bonusRolls).toEqual(1);
+    });
+    it("decreases by 1 after another roll", function() {
+      frame.roll(10);
+      frameTwo.roll(1);
+      expect(frame.bonusRolls).toEqual(1);
+    });
+  });
+
+  describe(".expendBonus()", function() {
+    it("reduces bonusRolls by 1", function() {
+      frame.roll(10);
+      frame.expendBonus(5);
+      expect(frame.bonusRolls).toEqual(1);
+    });
+    it("adds value to total", function() {
+      frame.roll(10);
+      frame.expendBonus(5);
+      expect(frame.total).toEqual(15);
+    });
+  });
+
+  describe(".accepts()", function() {
+    it("returns true if value doesn't take total above 10", function() {
+      expect(frame.accepts(5)).toEqual(true);
+    });
+    it("returns false if value greater than 10", function() {
+      expect(frame.accepts(11)).toEqual(false);
+    });
+    it("returns false if value less than 0", function() {
+      expect(frame.accepts(-1)).toEqual(false);
+    });
+    it("returns false if value is not an integer", function() {
+      expect(frame.accepts("string")).toEqual(false);
+    });
```

Comment:
This is a nice form of defensive type-checking. It might interest you to know that some languages introduce the idea of Types, which you can use to do this sort of thing more formally. For instance, Elm would allow you to write a `Score` type with valid values, that you could then use in the `Frame` instead of trying to exhaustively check across all sorts of things. Look into it!